### ISSUE TITLE
CPU training stalls for 60s+ compiling native artifact for Pool with batch normalization and quantized activation on backend cpu

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#278]
Public `recipe.model()
.pool(2).leak().norm(batch).iq(4).nl
.pool(2).tanh().iq(2).s
.loss(rmse)` stalls at 60 s on `backend=cpu`. Earliest failing operation is `.pool(2).leak().norm(batch).iq(4).nl`: it lowers through `lower_pool` (`Primitive::Pool`) combined with `.norm(batch)` (`Primitive::Normalize`, batch mode) and `.leak()` (`Activation::Leak` scalar `Select`/`Multiply` with 0.01 slope), plus `.iq(4).nl` and `.iq(2).s` integer quantization encoding. The dataset (`data/numeric/single_csv.csv`, ~630 k rows, 1 feature) and simple arithmetic (two pool stages, no parameters except quantization, loss=`rmse`) justify a sub-10 s epoch on CPU with 32 threads; measured 60 s stable exceeds workload and hardware justification. The `.pool(2).tanh().iq(2).s` stage is downstream and does not alter the earliest cause. No existing open issue covers this `Pool`/`Normalize`/quantized-activation native compilation path.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `6b480c8a8d84456d`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "CPU training stalls for 60s+ compiling native artifact for Pool with batch normalization and quantized activation on backend cpu",
  "body": "Public `recipe.model()\n.pool(2).leak().norm(batch).iq(4).nl\n.pool(2).tanh().iq(2).s\n.loss(rmse)` stalls at 60 s on `backend=cpu`. Earliest failing operation is `.pool(2).leak().norm(batch).iq(4).nl`: it lowers through `lower_pool` (`Primitive::Pool`) combined with `.norm(batch)` (`Primitive::Normalize`, batch mode) and `.leak()` (`Activation::Leak` scalar `Select`/`Multiply` with 0.01 slope), plus `.iq(4).nl` and `.iq(2).s` integer quantization encoding. The dataset (`data/numeric/single_csv.csv`, ~630 k rows, 1 feature) and simple arithmetic (two pool stages, no parameters except quantization, loss=`rmse`) justify a sub-10 s epoch on CPU with 32 threads; measured 60 s stable exceeds workload and hardware justification. The `.pool(2).tanh().iq(2).s` stage is downstream and does not alter the earliest cause. No existing open issue covers this `Pool`/`Normalize`/quantized-activation native compilation path.",
  "rationale": "Inspected `recipe.rs` (`lower_pool` at 5305 pushes `Primitive::Pool`, `.norm(batch)` pushes `Primitive::Normalize`, `.leak()` lowers to scalar `Activation::Leak` with slope parameter 0.01, and quantization `.iq(4).nl` / `.iq(2).s` encodes integer quantized formats). Read all open performance issues (253 MoE, 254 Bayes/perceptron, 252 CatBoost, 255 Forest/XGBoost, 277 LightGBM, 164 Scan/GRU, 171 Attention/Estimator, 250 kmeans, 271 NVIDIA nv0); none mention `Pool` or `Pool`+`Normalize`+quantization. The packet reaches the same earliest operation through the same `Pool`+`Normalize`+quantization native compilation path, not an existing issue. The workload size (~630 k rows, 1 feature) and CPU capabilities justify a sub-10 s epoch, so the 60 s stable measurement is materially slower than justified, confirming a distinct earliest cause requiring `new`."
}
```

## Deterministic failure packet

kind=performance
id=6b480c8a8d84456d
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:11076 next:11077 composition:38716026233484
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-130-11076/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/numeric/single_csv.csv")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.pool(2).leak().norm(batch).iq(4).nl
		.pool(2).tanh().iq(2).s
		.loss(rmse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(38716026233484)
		.epochs(1)
		.log(all)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #414.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #414: Verify the resolved CPU Pool compile path</summary>

Fixes #278

## Result

Current minimal completes the CPU Pool, batch-normalization, and quantized-block composition that previously exceeded 60 seconds. The merged native compile-preparation changes remove the historical stall, so this verification PR intentionally adds no production change.

The source CSV contains 16,000 complete data rows and 614,669 blank lines. The issue's approximate 630,000-row estimate counted blank lines; the public loader correctly trains on 12,800 rows after the 0.8 split.

## Public verification

- Built the release Recipe CLI from current minimal.
- Ran the exact composition through the public script interface with `RECIPE_FORCE_CPU=1` and a 65-second external limit.
- Removed its three native-cache entries before the measured cold run.
- Confirmed three cache misses and Clang 22.1.8 invocations with `-O2`.
- Measured 0.56 seconds wall time, 152,332 KiB maximum RSS, and a 1.487 ms epoch.
- Completed training with finite RMSE 37192.2969.

## Reachable issue paths

- Pool followed by XGBoost: 1.08 seconds, 5,760 rows, and finite loss.
- Pool followed by GRU: 6.77 seconds cold, 24,000 rows, and finite loss.
- Pool followed by LSTM: 2.38 seconds, 40,004 rows, and finite loss.

## Rejected patch

Changing only `pool_forward_body` from `alwaysinline` to `noinline` did not improve current minimal. The exact cold run changed from 0.56 to 0.58 seconds, the warm run changed from 0.24 to 0.25 seconds, and the cold GRU path changed from 6.77 to 6.78 seconds. Losses remained identical, so that unmerged patch is not included.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
